Cache pointer to function created by cling

Dr15Jones opened the pull request:
VTune showed that the previous way of calling into cling was causing
the ROOT internal mutex to be taken for each call. Philippe Canal
devised a way to access the cling generated function pointer directly
which avoids the locks.

This change required an update to ROOT 6.02 in order to work. That change is now in CMSSW_7_6.

Dr15Jones committed:
Cache pointer to function created by cling

VTune showed that the previous way of calling into cling was causing
the ROOT internal mutex to be taken for each call. Philippe Canal
devised a way to access the cling generated function pointer directly
which avoids the locks.

cmsbuild commented:
A new Pull Request was created by @Dr15Jones (Chris Jones) for CMSSW_7_6_X.

Cache pointer to function created by cling

It involves the following packages:

FWCore/Utilities


@cmsbuild, @smuzaffar, @Dr15Jones can you please review it and eventually sign? Thanks.
@makortel, @Martin-Grunewald, @wddgit, @wmtan this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@degano you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

Dr15Jones commented:
please test

Dr15Jones commented:
+1

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_6_X IBs after it passes the integration tests. This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_6_X IBs after it passes the integration tests. This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

cmsbuild commented:
-1 
 
Tested at: 6b87623a5bfe706bf0c4840e79cd4dcc3104267c
I found errors in the following unit tests: 
 
 ---> test runtestPhysicsToolsPatAlgos had ERRORS
 
 
you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11469/8404/summary.html

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_6_X IBs (but tests are reportedly failing). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11469/8404/summary.html

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_6_X IBs (but tests are reportedly failing). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_6_X IBs after it passes the integration tests. This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

Dr15Jones commented:
As far as I can tell, this unit test failure is independent of my code changes. The test fails in the new ```ExpressionEvaluator``` which I believe does not use the code I changed. I will double check to be sure.

Dr15Jones commented:
please check

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_6_X IBs after it passes the integration tests. This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

Dr15Jones commented:
I double checked and the failing unit test doesn't make use of any of the code from this pull request.

cmsbuild commented:
-1 
 
Tested at: 6b87623a5bfe706bf0c4840e79cd4dcc3104267c
I found errors in the following unit tests: 
 
 ---> test testRecoMETMETProducers had ERRORS
---> test runtestPhysicsToolsPatAlgos had ERRORS
 
 
you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11469/8408/summary.html

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_6_X IBs (but tests are reportedly failing). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

Dr15Jones commented:
The second failure in the new test is failing in the IB now.

Dr15Jones commented:
runtestPhysicsToolsPatAlgos fails due to a problem with a ROOT dictionary while openning a file
```
26-Sep-2015 18:42:10 CEST  Initiating request to open file root://eoscms//eos/cms/store/relval/CMSSW_7_6_0_pre4/RelValProdTTbar_13/AODSIM/76X_mcRun2_asymptotic_v1-v1/00000/90C5FA7A-7B4F-E511-BC27-002618943916.root?svcClass=default
%MSG-w XrdAdaptor:  file_open 26-Sep-2015 18:42:10 CEST pre-events
Data is served from cern.ch instead of original site eoscms
%MSG
26-Sep-2015 18:42:11 CEST  Successfully opened file root://eoscms//eos/cms/store/relval/CMSSW_7_6_0_pre4/RelValProdTTbar_13/AODSIM/76X_mcRun2_asymptotic_v1-v1/00000/90C5FA7A-7B4F-E511-BC27-002618943916.root?svcClass=default
input_line_428:85:26: error: no member named 'TriggerObjectType' in namespace 'trigger'
    std::vector<trigger::TriggerObjectType> v_t_tot;
                ~~~~~~~~~^
input_line_428:86:26: error: no member named 'TriggerObjectType' in namespace 'trigger'
    std::vector<trigger::TriggerObjectType>::const_iterator v_t_tot_ci;
                ~~~~~~~~~^
input_line_428:86:46: error: non-friend class member 'const_iterator' cannot have a qualified name
    std::vector<trigger::TriggerObjectType>::const_iterator v_t_tot_ci;
                                           ~~^
input_line_428:86:60: error: expected ';' at end of declaration list
    std::vector<trigger::TriggerObjectType>::const_iterator v_t_tot_ci;
                                                           ^
                                                           ;
input_line_428:87:39: error: no member named 'TriggerObjectType' in namespace 'trigger'
    edm::Wrapper<std::vector<trigger::TriggerObjectType> > w_v_t_tot;
                             ~~~~~~~~~^
input_line_428:87:58: error: expected a type
    edm::Wrapper<std::vector<trigger::TriggerObjectType> > w_v_t_tot;
                                                         ^
26-Sep-2015 18:42:20 CEST  Closed file root://eoscms//eos/cms/store/relval/CMSSW_7_6_0_pre4/RelValProdTTbar_13/AODSIM/76X_mcRun2_asymptotic_v1-v1/00000/90C5FA7A-7B4F-E511-BC27-002618943916.root?svcClass=default
----- Begin Fatal Exception 26-Sep-2015 18:42:20 CEST-----------------------
An exception of category 'FatalRootError' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=PATTriggerMatcherDRDPtLessByR label='somePatJetTriggerMatchHLTMu8DiJet30'
   Additional Info:
      [a] Fatal Root Error: @SUB=TInterpreter::AutoParse
Error parsing payload code for class trigger::TriggerObjectType with content:
[the copy of DataFormats/HLTReco/src/classes.h]
```

While the other test testRecoMETMETProducers can't find a file

```
----- Begin Fatal Exception 26-Sep-2015 18:27:48 CEST-----------------------
An exception of category 'FileOpenError' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing input source of type PoolSource
   [2] Calling RootFileSequenceBase::initTheFile()
   [3] Calling StorageFactory::open()
   [4] Calling File::sysopen()
Exception Message:
Failed to open the file 'ttbarForMetTests.root'
   Additional Info:
      [a] Input file file:ttbarForMetTests.root could not be opened.
      [b] open() failed with system error 'No such file or directory' (error code 2)
----- End Fatal Exception -------------------------------------------------
Failure using recoMET_genMet_cfg.py: status 84
```

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11469/8408/summary.html

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_6_X IBs (but tests are reportedly failing). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

Dr15Jones commented:
I think I understand the problem with runtestPhysicsToolsPatAlgos. The cut parser was told to apply cuts to the type ```trigger::TriggerObjectType``` but that type is an enum. Therefore when the cut parser attempts to find 'methods' (which it now does right up front) this fails.

Dr15Jones committed:
Added included needed for cling during cut parser evaluation

ROOT's cling was failing while the cut parser was evaluating an
expression because a type was not found. Adding the header explicitly
to the classes.h solved the problem.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #11469 was updated. @perrotta, @smuzaffar, @cmsbuild, @Martin-Grunewald, @Dr15Jones, @fwyzard can you please check and sign again.

Dr15Jones commented:
+1

cmsbuild commented:
-1 
 
Tested at: 5b2dcc45e35bf65027aa0cc4246abc8b0d906dcf
I found errors in the following unit tests: 
 
 ---> test testRecoMETMETProducers had ERRORS
 
 
you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11469/8414/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11469/8414/summary.html

Dr15Jones commented:
This same test has already been failing in the IBs
https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc6_amd64_gcc493/CMSSW_7_6_X_2015-09-24-2300/unitTestLogs/RecoMET/METProducers

Martin-Grunewald commented:
please test

Martin-Grunewald commented:
+1

Dr15Jones commented:
@Martin-Grunewald The tests won't succeed since the failure is in the IB

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_6_X IBs after it passes the integration tests. This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_6_X IBs after it passes the integration tests. This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

Dr15Jones commented:
@smuzaffar when the tests are run, is this pull request merged with the head of CMSSW_7_6_X or is it run as is? I'm trying to determine how the failure from the IB is being tickled by this pull request.

smuzaffar commented:
From the PR test job I see the following
```+ git cms-merge-topic -u 11469
No release tags specified, using default CMSSW_7_6_X_2015-09-27-2300.
You are on branch CMSSW_7_6_X
Checking out FWCore/Version in tag CMSSW_7_6_X_2015-09-27-2300.
Cloning into '/tmp/cmsbuild/workspace/ib-any-integration/CMSSW_7_6_X_2015-09-27-2300/src'...
Switched to a new branch 'from-CMSSW_7_6_X_2015-09-27-2300'
Branch from-CMSSW_7_6_X_2015-09-27-2300 set up to track remote branch CMSSW_7_6_X from official-cmssw.
HEAD is now at 2aad5e3 Merge pull request #11500 from davidlt/resolve-private-protected-part1
From .
 * [new ref]                    -> merge-attempt
From https://github.com/cms-sw/cmssw
 * [new ref]         refs/pull/11469/head -> cms-sw/refs/pull/11469/head
Switched to branch 'merge-attempt'
Merge made by the 'recursive' strategy.
 DataFormats/HLTReco/src/classes.h             |  1 +
 FWCore/Utilities/interface/FunctionWithDict.h |  3 +++
 FWCore/Utilities/src/FunctionWithDict.cc      | 21 ++++++++++++++-------
 3 files changed, 18 insertions(+), 7 deletions(-)
Switched to branch 'from-CMSSW_7_6_X_2015-09-27-2300'
Your branch is behind 'official-cmssw/CMSSW_7_6_X' by 32 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)```

so to me it looks like it is run as it is i.e. without merging it to the head of IB branch e.g. CMSSW_7_6_X

Dr15Jones commented:
That would account for the unit test failure then. The place where my code 'branched' from CMSSW_7_6_X, that version of CMSSW had the version of the code that caused the unrelated test failure. Since ```git cms-checkdeps -a``` will get the version from my branch and not the release the unit test will fail. But when my changes are merged into CMSSW_7_6_X the problem will go away.

cmsbuild commented:
+1
Tested at: 5b2dcc45e35bf65027aa0cc4246abc8b0d906dcf
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11469/8448/summary.html

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_6_X IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11469/8448/summary.html

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_6_X IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

davidlange6 commented:
+1
